[RBD Async DR] Fix resize/Delete volume when it moved to another cluster

Madhu-1 opened the pull request:
In the case of the Async DR, the volumeID will not be the same if the clusterID or the PoolID is different, With Earlier implementation, it is expected that the new volumeID mapping is stored in the rados omap pool. In the case of the
ControllerExpand or the DeleteVolume Request, the only volumeID will be sent it's not possible to find the corresponding poolID in the new cluster.

With This Change, it works as below

The csi-rbdplugin-controller will watch for the PV objects, when there are any PV objects created it will check the omap already exists, If the omap doesn't exist it will generate the new volumeID and it checks for the volumeID mapping entry in the PV annotation, if the mapping does not exist, it will add the new entry to the PV annotation.

The cephcsi will check for the PV annotations if the omap does not exist if the mapping exists in the PV annotation, it will use the new volumeID for further operations.

fixes: #1936
closes: #1938 (this is an alternative)
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd:store/Read volumeID in/from PV annotation

In the case of the Async DR, the volumeID will
not be the same if the clusterID or the PoolID
is different, With Earlier implementation, it
is expected that the new volumeID mapping is
stored in the rados omap pool. In the case of the
ControllerExpand or the DeleteVolume Request,
the only volumeID will be sent it's not possible
to find the corresponding poolID in the new cluster.

With This Change, it works as below

The csi-rbdplugin-controller will watch for the PV
objects, when there are any PV objects created it
will check the omap already exists, If the omap doesn't
exist it will generate the new volumeID and it checks for
the volumeID mapping entry in the PV annotation, if the
mapping does not exist, it will add the new entry
to the PV annotation.

The cephcsi will check for the PV annotations if the
omap does not exist if the mapping exists in the PV
annotation, it will use the new volumeID for further
operations.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@nixpanic as discussed one more approach other than  https://github.com/ceph/ceph-csi/pull/1938 to fix this issue

nixpanic commented on internal/rbd/rbd_journal.go:
```diff
@@ -25,6 +25,13 @@ import (
 
+const (
+	// PVVolumeHandleLabel is the annotation key set on the PV object.
+	PVVolumeHandleLabel = "CephCSIVolumeHandle"
```
Labels are commonly formatted like `csi.ceph.io/volume-handle`. All lowercase, with a prefix (`csi.ceph.io/`) for the application/component, and sub-label (`volume-handle`) as real 'label'.

It would be good to adapt such a format here too.

Madhu-1 replied:
This is for annotation sorry need to change the name. is it suggested to use `csi.ceph.io/volume-handle` for annotation too?

nixpanic replied:
Yes, I think so. Identifying the application/component that set a Label or Annotation seems to be good practice.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -881,6 +867,31 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
+// the structure with elements from on-disk image metadata as well.
+func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+	vol, err := generateVolumeFromVolumeID(ctx, volumeID, cr, secrets)
+	if errors.Is(err, util.ErrKeyNotFound) || errors.Is(err, util.ErrPoolNotFound) {
```
Easier to do the `if` the other way around. A small body inside the statement returning `nil, err`, and continue with the rest outside the if-statement like it is the default behaviour.

When an error occurs, ideally return `nil` and an error, not some incomplete `vol` object. Make sure to call `vol.Destroy()` in error cases too.

Madhu-1 replied:
Need to return vol even in case of error to destroy the context and to release the lock

nixpanic replied:
really? where is the lock taken then? Ideally it should be released at the same level...

Madhu-1 replied:
the lock was taken on the volumeID when and common function is added to cleanup it https://github.com/ceph/ceph-csi/pull/1924/files#diff-3792fcc7313c4d1e0b81c4d3ff7ac714b9d299130641673729196b2a4a1d6f2aR175-R186. if the PR is merged I will rework that one. Will fix it in this PR.

nixpanic commented on internal/rbd/rbd_journal.go:
```diff
@@ -496,7 +502,7 @@ func RegenerateJournal(imageName, volumeID, pool, journalPool, requestName strin
 	err := vi.DecomposeCSIID(rbdVol.VolID)
 	if err != nil {
-		return fmt.Errorf("%w: error decoding volume ID (%s) (%s)",
+		return rbdVol.VolID, fmt.Errorf("%w: error decoding volume ID (%s) (%s)",
```
In error cases, it is better to return an empty string.

nixpanic commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -122,6 +122,40 @@ func checkStaticVolume(pv *corev1.PersistentVolume) (bool, error) {
+	if replicationLabelExists && volumeIDExistInAnnotation {
+		return nil
+	}
+	pv.Labels[rbd.PVReplicatedLabel] = "true"
```
I'm not a big fan of `true` strings, maybe some status like `volume-handle-detected`? Just a nit, not required to change as the value is not *that* important.

nixpanic reviewed (commented):
I prefer this approach much more over #1936, thanks for posting this version!

At the moment you're watching for PVs, I hope this can be moved towards the VolumeReplication CR at a later time. From there you can then get the PVC and PV too.

nixpanic commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -122,6 +122,40 @@ func checkStaticVolume(pv *corev1.PersistentVolume) (bool, error) {
 
+// storeVolumeIDInPV stores the new volumeID in PV object.
+func (r ReconcilePersistentVolume) storeVolumeIDInPV(pv *corev1.PersistentVolume, newVolumeID string) error {
+	// take a lock to avoid concurrent update of configmap
```
not sure what this comment means

Madhu-1 replied:
Leftover removing it

nixpanic commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -122,6 +122,40 @@ func checkStaticVolume(pv *corev1.PersistentVolume) (bool, error) {
 }
 
+// storeVolumeIDInPV stores the new volumeID in PV object.
+func (r ReconcilePersistentVolume) storeVolumeIDInPV(pv *corev1.PersistentVolume, newVolumeID string) error {
```
There probably should be a verification that the PV is a Ceph-CSI one?

Madhu-1 replied:
The check already exists. the controller will reconcile the only PV created for this Driver.

agarwal-mudit commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -122,6 +122,39 @@ func checkStaticVolume(pv *corev1.PersistentVolume) (bool, error) {
+// storeVolumeIDInPV stores the new volumeID in PV object.
+func (r ReconcilePersistentVolume) storeVolumeIDInPV(pv *corev1.PersistentVolume, newVolumeID string) error {
+	if v, ok := pv.Annotations[rbd.PVVolumeHandleAnnotationKey]; ok {
+		if v == newVolumeID {
```
if this is true, why will it go to Line 141?

Madhu-1 replied:
as its already existing in annotation, it will get return at line 129

agarwal-mudit replied:
yeah, then why do we need to check it again in 141

nixpanic commented on internal/rbd/rbd_journal.go:
```diff
@@ -25,6 +25,15 @@ import (
+	// PVVolumeHandleAnnotationKey is the annotation key set on the PV object.
+	PVVolumeHandleAnnotationKey = "csi.ceph.io/volume-handle"
+	// PVReplicatedLabelKey is the label key set on PV object.
+	PVReplicatedLabelKey = "CephCSIReplicatedVolume"
```
Same as with annotations, use `csi.ceph.io/repliacted-volume` to follow the naming other projects do

Madhu-1 replied:
done.

ShyamsundarR commented:
Annotations are better than config maps.

@Madhu-1 and me discussed an alternate that provides configuration to the CSI driver for the required mapping, than annotate PVs. Madhu is thinking through the same, and if we discuss and find that a better long term alternative, we would possibly got that route.

Till then, without hopefully confusing matters further, this is an acceptable stand-in for the problem being resolved.

nixpanic reviewed (dismissed):
CI failures happened due to timeouts.

Just use `@mergifyio rebase` to re-test once a 2nd approval is in.

Madhu-1 commented:
@humblec PTAL

humblec commented:
> @humblec PTAL

@Madhu-1 do we have any design doc (https://github.com/ceph/ceph-csi/pull/1938#issuecomment-810304643) for this async work in general ( may be one which cover the integration to the other component too ?) apart from the one available here https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-mirror.md for reference ? the one available here does not carry much information and I am finding difficulty to get the workflow , so the ask.

Madhu-1 commented:
> > @humblec PTAL
> 
> @Madhu-1 do we have any design doc ([#1938 (comment)](https://github.com/ceph/ceph-csi/pull/1938#issuecomment-810304643)) for this async work in general ( may be one which cover the integration to the other component too ?) apart from the one available here https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-mirror.md for reference ? the one available here does not carry much information and I am finding difficulty to get the workflow , so the ask.

@humblec as i  already explain the currently planned design for you and its clear, this is the doc which I have created initially https://hackmd.io/1Pbj_9rzQ2SwK80GK6-qWg?both. i don't have the LLD

humblec commented:
> > > @humblec PTAL
> > 
> > 
> > @Madhu-1 do we have any design doc ([#1938 (comment)](https://github.com/ceph/ceph-csi/pull/1938#issuecomment-810304643)) for this async work in general ( may be one which cover the integration to the other component too ?) apart from the one available here https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-mirror.md for reference ? the one available here does not carry much information and I am finding difficulty to get the workflow , so the ask.
> 
> @humblec as i already explain the currently planned design for you and its clear, this is the doc which I have created initially https://hackmd.io/1Pbj_9rzQ2SwK80GK6-qWg?both. i don't have the LLD

@Madhu-1 looks like many changes  from the initial thoughts keep happening  and was struggling to keep updated on those. Thanks for the link. Hopefully it helps :) .

>https://github.com/ceph/ceph-csi/pull/1946#issuecomment-811115845
>

Are we saying the annotation method ( this PR) will be kept as a `temporary` one ?  if yes, what  are we planning as a permanent solution mentioned by above ?

Madhu-1 commented:
> > > > @humblec PTAL
> > > 
> > > 
> > > @Madhu-1 do we have any design doc ([#1938 (comment)](https://github.com/ceph/ceph-csi/pull/1938#issuecomment-810304643)) for this async work in general ( may be one which cover the integration to the other component too ?) apart from the one available here https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-mirror.md for reference ? the one available here does not carry much information and I am finding difficulty to get the workflow , so the ask.
> > 
> > 
> > @humblec as i already explain the currently planned design for you and its clear, this is the doc which I have created initially https://hackmd.io/1Pbj_9rzQ2SwK80GK6-qWg?both. i don't have the LLD
> 
> @Madhu-1 looks like many changes from the initial thoughts keep happening and was struggling to keep updated on those. Thanks for the link. Hopefully it helps :) .
> 
> > [#1946 (comment)](https://github.com/ceph/ceph-csi/pull/1946#issuecomment-811115845)
> 
> Are we saying the annotation method ( this PR) will be kept as a `temporary` one ? if yes, what are we planning as a permanent solution mentioned by above ?

Yes am working on a new design need to check the feasibility, it takes some time (not soon) to get the design doc out.

humblec commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -161,11 +179,18 @@ func (r ReconcilePersistentVolume) reconcilePV(obj runtime.Object) error {
 		}
 		defer cr.DeleteCredentials()
 
-		err = rbd.RegenerateJournal(imageName, volumeHandler, pool, journalPool, requestName, cr)
```
@Madhu-1 we call `regeneratejournal` without any locking in place, so while reconciling a PV, if regenrateJournal() was called more than once, Isnt there a possibility of a race scenario ?

humblec commented on internal/controller/persistentvolume/persistentvolume.go:
```diff
@@ -161,11 +179,18 @@ func (r ReconcilePersistentVolume) reconcilePV(obj runtime.Object) error {
 			return err
 		}
+		if rbdVolID != volumeHandler {
+			err = r.storeVolumeIDInPV(pv, rbdVolID)
```
imo, the storeVolumeIDinPV should be a protected call. Otherwise the other parallel request may see a inconsistency in some scenarios.

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -881,6 +867,32 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
+		if pvlist.Items[i].Spec.CSI != nil && pvlist.Items[i].Spec.CSI.VolumeHandle == volumeID {
+			if v, ok := pvlist.Items[i].Annotations[PVVolumeHandleAnnotationKey]; ok {
+				util.UsefulLog(ctx, "found new volumeID %s for existing volumeID %s", v, volumeID)
+				return generateVolumeFromVolumeID(ctx, v, cr, secrets)
```
@Madhu-1 this is bit confusing. That said, Is it a problem if we see new volume ID for existing Volume ID?  can we add some source code comment about this block logic ?

Madhu-1 replied:
@humblec it's not a problem. add more comments

Madhu-1 committed:
rbd: take lock when reconciling the PV

there can be a change we can reconcile same
PV parallelly we can endup in generating and
deleting multiple omap keys. to be on safer
side taking lock to process one volumeHandle
at a time.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

humblec commented:
@Madhu-1 looks like latest patchset address https://github.com/ceph/ceph-csi/pull/1946#discussion_r607932043,  is this ready for review or you are still polishing it ?

Madhu-1 commented:
> @Madhu-1 looks like latest patchset address [#1946 (comment)](https://github.com/ceph/ceph-csi/pull/1946#discussion_r607932043), is this ready for review or you are still polishing it ?

This is ready for review

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -883,6 +869,38 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
+
+	// If the volume details are not found in the OMAP it can be a mirrored RBD
+	// image and the OMAP is already generated and the volumeHandle might not
+	// be the same in the PV.Spec.CSI.VolumeHandle. Check the PV annotation for
```
if the volume details are not in OMAP , it could also be any static PV customer created in that cluster ( without mirroring), Isnt it ? why I am asking this is to confirm, we wont confuse with static PVs created by the admin and the mirrored PVs and thus land into performing different operations. or we are good to distinguish the PVs based on the key set by replication operator just for mirroring ?

Madhu-1 replied:
AFAIC this won't be called for static PVC. and we don't support any operator for static PVC except mounting and unmounting?

ShyamsundarR replied:
Also, due to the label selector, the static PV created by users would be exempt from the selection. This should not be an issue.

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -883,6 +869,38 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
+	listOpt := metav1.ListOptions{
+		LabelSelector: PVReplicatedLabelKey,
+	}
+	pvlist, pErr := c.CoreV1().PersistentVolumes().List(context.TODO(), listOpt)
```
This is a heavy operation, that said, listing out all the PVs in the cluster and fetching the volume handle looks to cause performance issues especially in large clusters,.

Madhu-1 replied:
This will be fetched only if the omap is not found and we are fetching based on the selector. as this is a temp solution i don't see a problem with it.

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -777,9 +778,8 @@ func genSnapFromSnapID(ctx context.Context, rbdSnap *rbdSnapshot, snapshotID str
-// the structure with elements from on-disk image metadata as well.
-func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
+// generateVolumeFromVolumeID generates a rbdVolume structure from the provided identifier.
+func generateVolumeFromVolumeID(ctx context.Context, volumeID string, cr *util.Credentials, secrets map[string]string) (*rbdVolume, error) {
```
nit: compared to previous scenarios, with the introduction of mirroring, iic, there is a possibility of volumeID being `nil`. If yes, DecompoeCSIID can be  avoided at start of this function if volumeID is nil.

Madhu-1 replied:
@humblec can you please suggest in which scenario the volumeId can be nil?

ShyamsundarR replied:
Did a cursory check for `nil` possibility, did not find any.

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -883,6 +869,38 @@ func genVolFromVolID(ctx context.Context, volumeID string, cr *util.Credentials,
+		return vol, pErr
+	}
+	for i := range pvlist.Items {
+		if pvlist.Items[i].Spec.CSI != nil && pvlist.Items[i].Spec.CSI.VolumeHandle == volumeID {
```
if the listOpt is `PVReplicatedLabelKey` why do we want to check whether its a CSI spec below?.

Madhu-1 replied:
It's safer check to check for nil pointer before accessing it.
